email/parser.go

```diff
@@ -0,0 +1,281 @@
+package email
+
+import (
+	"abuse-scanner/database"
+	"bufio"
+	"context"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+const (
+	// parseFrequency defines the frequency with which the parser looks for
+	// emails to be parsed
+	parseFrequency = 25 * time.Second
+)
+
+var (
+	skylinkRE           = regexp.MustCompile("^.*([a-z0-9]{55})|([a-zA-Z0-9-_]{46}).*$")
+	validateSkylink64RE = regexp.MustCompile("^([a-zA-Z0-9-_]{46})$")
+	validateSkylink32RE = regexp.MustCompile("^([a-zA-Z0-9-_]{55})$")
+)
+
+type (
+	// Parser is an object that will periodically scan for unparsed emails and
+	// parse them for skylinks.
+	Parser struct {
+		staticContext   context.Context
+		staticDatabase  *database.AbuseScannerDB
+		staticLogger    *logrus.Entry
+		staticSponsor   string
+		staticWaitGroup sync.WaitGroup
+	}
+)
+
+// NewParser creates a new parser.
+func NewParser(ctx context.Context, database *database.AbuseScannerDB, sponsor string, logger *logrus.Logger) *Parser {
+	return &Parser{
+		staticContext:  ctx,
+		staticDatabase: database,
+		staticLogger:   logger.WithField("module", "Parser"),
+		staticSponsor:  sponsor,
+	}
+}
+
+// Start initializes the fetch process.
+func (p *Parser) Start() error {
+	p.staticWaitGroup.Add(1)
+	go func() {
+		p.threadedParseMessages()
+		p.staticWaitGroup.Done()
+	}()
+	return nil
+}
+
+// Stop waits for the parser's waitgroup and times out after one minute.
+func (p *Parser) Stop() error {
```

Comment:
Seems like you got some hybrid solution now.

`Stop` doesn't stop anything. It just waits and you manually close the context somewhere else.
But instead, this should close the `Parser`'s context first in here and then wait.

Either way this is not great because waiting on a waitgroup in a goroutine like that isn't clear but we are under a bit of a time constraint so let's f/u with that.

What we want to do in the f/u is to make this shutdown a 2 step process. 

1. set some flag (e.g. close a shutdown channel) to signal that it shouldn't start a new job
2. wait for ongoing jobs to be done
3. return `nil` if they finished, otherwise interrupt the ongoing jobs and return an error